File: tools/sentences.py
"""THE sentence rule, Python side (graph/sentence-vectors.json).

The mirror of graph/sentences.ts. Two surfaces must agree on where a sentence
ends — the viewer, which knows which sentence was clicked, and the backend,
which applies the decision to the author's file — so the rule is specified once
and both languages are held to it by the shared vectors.

Worth stating plainly, as arc-core does elsewhere when a thing looks unused:
no Python caller needs this today. arc's backend is TypeScript, and validate.py
has no business splitting prose. It exists because a rule proven in one
language is a rule that can fork in the other, and the date rule earned that
lesson already. When a Python consumer arrives — a manuscript importer, a
prose-side check — the rule is here and is provably the same one.

THE BIAS: when unsure, do not split. An unsplit pair costs a coarser judgment;
a wrong split costs half a sentence landing in the author's book.
"""
from __future__ import annotations

import re
from typing import NamedTuple


class Sentence(NamedTuple):
    """The exact substring, plus where it came from."""
    text: str
    start: int
    end: int   # exclusive


# Words ending in a period without ending a sentence. Lowercased, no dot.
ABBREVIATIONS = {
    "mr", "mrs", "ms", "dr", "prof", "st", "sr", "jr", "rev", "fr",
    "gen", "col", "capt", "lt", "sgt", "maj", "cpl", "adm",
    "mt", "ave", "blvd", "rd", "no", "vol", "pp", "ed",
    "inc", "ltd", "co", "corp", "dept", "univ",
    "etc", "vs", "al", "cf", "approx", "est",
}

CLOSERS = {'"', "'", ")", "]", "}", "”", "’", "»"}

_WORD_CHAR = re.compile(r"[A-Za-z.]")


def _at(text: str, i: int) -> str | None:
    return text[i] if 0 <= i < len(text) else None


def _is_digit(c: str | None) -> bool:
    return c is not None and c.isdigit() and c.isascii()


def _is_space(c: str | None) -> bool:
    return c is not None and c.isspace()


def _is_lower(c: str | None) -> bool:
    """A letter with a distinct upper form — used only to spot a speech tag."""
    return c is not None and c != c.upper() and c == c.lower()


def _word_before(text: str, i: int) -> str:
    j = i
    while j > 0 and _WORD_CHAR.match(text[j - 1]):
        j -= 1
    return text[j:i]


def _terminates_at(text: str, i: int) -> bool:
    c = text[i]

    # An ellipsis trails off; it does not terminate.
    if c == "." and (_at(text, i + 1) == "." or _at(text, i - 1) == "."):
        return False
    if _at(text, i - 1) == "…" or c == "…":
        return False

    if c == ".":
        # 3.14 — a period between digits is never a boundary.
        if _is_digit(_at(text, i - 1)) and _is_digit(_at(text, i + 1)):
            return False
        word = _word_before(text, i)
        # A single letter is an initial: J. R. R. Tolkien.
        if len(word) == 1 and word.isalpha():
            return False
        if word.lower().replace(".", "") in ABBREVIATIONS:
            return False

    # Step past closing quotes and brackets.
    j = i + 1
    while j < len(text) and text[j] in CLOSERS:
        j += 1

    if j >= len(text):
        return True
    if not _is_space(_at(text, j)):
        return False

    # `"Get out!" he said.` — a lowercase word after the mark is a speech tag.
    k = j
    while k < len(text) and _is_space(text[k]):
        k += 1
    if k >= len(text):
        return True
    return not _is_lower(_at(text, k))


def split_sentences(paragraph: str) -> list[Sentence]:
    """Split a paragraph into sentences that TILE it.

    Contiguous, gapless, covering the paragraph end to end — trailing
    whitespace belongs to the sentence it follows, so
    ``"".join(s.text for s in split_sentences(p)) == p`` holds by construction.
    """
    if paragraph == "":
        return []
    out: list[Sentence] = []
    start = 0
    i = 0
    while i < len(paragraph):
        c = paragraph[i]
        if c not in (".", "!", "?", "…") or not _terminates_at(paragraph, i):
            i += 1
            continue

        end = i + 1
        while end < len(paragraph) and paragraph[end] in CLOSERS:
            end += 1
        while end < len(paragraph) and paragraph[end].isspace():
            end += 1

        out.append(Sentence(paragraph[start:end], start, end))
        start = end
        i = end

    if start < len(paragraph):
        out.append(Sentence(paragraph[start:], start, len(paragraph)))
    return out


def sentence_at(paragraph: str, offset: int) -> int:
    """Which sentence covers this offset? -1 when the paragraph has none."""
    sentences = split_sentences(paragraph)
    for idx, s in enumerate(sentences):
        if s.start <= offset < s.end:
            return idx
    return len(sentences) - 1 if sentences else -1

File: graph/sentences.ts
// THE sentence rule, specified once (graph/sentence-vectors.json).
//
// Two surfaces need to agree on where a sentence ends: the viewer, which must
// know which sentence the author right-clicked, and the backend, which applies
// the decision to the author's file. Two implementations of one rule with
// nothing forcing them to agree is a bug waiting for a specific sentence — and
// here the bug's shape is the wrong sentence being accepted into a book,
// silently. So the rule lives here, beside the date rule, and a shared vector
// file holds both languages to it.
//
// THE BIAS, STATED: when the rule is unsure, it does NOT split. An unsplit
// sentence costs the author a coarser judgment — two sentences taken or
// refused together, which is exactly what the paragraph-level verbs already
// offer. A wrong split costs half a sentence landing in the manuscript. The
// harms are not the same size, so the tie goes to under-splitting. Every
// judgment call below runs that direction.

/** One sentence, with the offsets it came from. Offsets, because a caller has
 *  to map a click position back to a sentence and put text where it belongs. */
export interface Sentence {
  /** The exact substring: paragraph.slice(start, end). */
  text: string
  start: number
  /** Exclusive. */
  end: number
}

/** Words that end in a period without ending a sentence. Lowercased, no dot.
 *  Deliberately short: a missing entry under-splits (safe), a wrong entry
 *  over-splits (not). */
const ABBREVIATIONS = new Set([
  'mr', 'mrs', 'ms', 'dr', 'prof', 'st', 'sr', 'jr', 'rev', 'fr',
  'gen', 'col', 'capt', 'lt', 'sgt', 'maj', 'cpl', 'adm',
  'mt', 'ave', 'blvd', 'rd', 'no', 'vol', 'pp', 'ed',
  'inc', 'ltd', 'co', 'corp', 'dept', 'univ',
  'etc', 'vs', 'al', 'cf', 'approx', 'est',
])

/** Closers that may sit between the terminal mark and the space: quotes and
 *  brackets, straight and curly. */
const CLOSERS = new Set(['"', "'", ')', ']', '}', '”', '’', '»'])

const isDigit = (c: string | undefined): boolean => c !== undefined && c >= '0' && c <= '9'
const isSpace = (c: string | undefined): boolean => c !== undefined && /\s/.test(c)

/** Lowercase for boundary purposes: a letter that has a distinct upper form.
 *  Used only to spot a continuation — `"Get out!" he said.` */
const isLower = (c: string | undefined): boolean =>
  c !== undefined && c !== c.toUpperCase() && c === c.toLowerCase()

/** The word ending at index `i` (exclusive), letters and dots only. */
function wordBefore(text: string, i: number): string {
  let j = i
  while (j > 0 && /[A-Za-z.]/.test(text[j - 1])) j--
  return text.slice(j, i)
}

/** Does a sentence end at `i`, where text[i] is a terminal mark? */
/** Is the mark at `i` a genuine terminal mark — not an ellipsis trailing
 *  off, a decimal point, an initial, or an abbreviation? The mark-level half
 *  of the sentence rule, exported for the checks that need the mark WITHOUT
 *  the follower rule: a lowercase word after a terminal period is a speech
 *  tag to the splitter and a typo to the mechanical checker, and both must
 *  agree on what a terminal mark IS or they are two sentence rules. */
export function isTerminalMark(text: string, i: number): boolean {
  const c = text[i]

  // An ellipsis trails off; it does not terminate. Under-splits by design.
  if (c === '.' && (text[i + 1] === '.' || text[i - 1] === '.')) return false
  if (text[i - 1] === '…' || c === '…') return false

  if (c === '.') {
    // 3.14, 1848.5 — a period between digits is never a boundary.
    if (isDigit(text[i - 1]) && isDigit(text[i + 1])) return false
    const word = wordBefore(text, i)
    // A single letter is an initial: J. R. R. Tolkien.
    if (word.length === 1 && /[A-Za-z]/.test(word)) return false
    // Mr. Reyes. Also catches e.g. and i.e., whose last segment is one letter,
    // via the initial rule above.
    if (ABBREVIATIONS.has(word.toLowerCase().replace(/\./g, ''))) return false
  }
  return true
}

function terminatesAt(text: string, i: number): boolean {
  if (!isTerminalMark(text, i)) return false

  // Step past closing quotes and brackets: the mark may belong to a quotation.
  let j = i + 1
  while (j < text.length && CLOSERS.has(text[j])) j++

  // End of paragraph always ends the sentence.
  if (j >= text.length) return true

  // A boundary needs whitespace after it. `Mr.Reyes` and `1.5` never reach here.
  if (!isSpace(text[j])) return false

  // What follows the space decides the quotation case. `"Get out!" he said.`
  // continues — a lowercase word after a terminal mark is a speech tag, not a
  // new sentence. A capital, a digit, or a quote starts one.
  let k = j
  while (k < text.length && isSpace(text[k])) k++
  if (k >= text.length) return true
  return !isLower(text[k])
}

/** Split a paragraph into sentences.
 *
 *  The sentences TILE the paragraph: they are contiguous, gapless, and cover
 *  it end to end, so trailing whitespace belongs to the sentence it follows.
 *  That is what makes `split(p).map(s => s.text).join('') === p` true by
 *  construction rather than by care — and that identity is the whole reason
 *  accepting one sentence cannot reflow its neighbours. */
export function splitSentences(paragraph: string): Sentence[] {
  if (paragraph === '') return []
  const out: Sentence[] = []
  let start = 0

  for (let i = 0; i < paragraph.length; i++) {
    const c = paragraph[i]
    if (c !== '.' && c !== '!' && c !== '?' && c !== '…') continue
    if (!terminatesAt(paragraph, i)) continue

    // Take the closers, then the whitespace that separates this sentence from
    // the next — it is this sentence's, so rejoining is lossless.
    let end = i + 1
    while (end < paragraph.length && CLOSERS.has(paragraph[end])) end++
    while (end < paragraph.length && isSpace(paragraph[end])) end++

    out.push({ text: paragraph.slice(start, end), start, end })
    start = end
    i = end - 1
  }

  if (start < paragraph.length) {
    out.push({ text: paragraph.slice(start), start, end: paragraph.length })
  }
  return out
}

/** Which sentence covers this offset? The click-to-sentence mapping, and the
 *  only supported way to name a sentence across the wire: an index into
 *  splitSentences(paragraph). Returns -1 when the paragraph has none. */
export function sentenceAt(paragraph: string, offset: number): number {
  const sentences = splitSentences(paragraph)
  for (let i = 0; i < sentences.length; i++) {
    if (offset >= sentences[i].start && offset < sentences[i].end) return i
  }
  return sentences.length ? sentences.length - 1 : -1
}
